src/environmentd/src/http/webhook.rs

```diff
@@ -56,14 +64,39 @@ pub async fn handle_webhook(
         body_ty,
         header_tys,
         validator,
-    } = client
-        .append_webhook(database, schema, name, conn_id, received_at)
-        .await?;
+    } = async {
+        let mut guard = webhook_cache.entries.lock().await;
+
+        // Remove the appender from our map, only re-insert it, if it's valid.
+        match guard.remove(&(database.clone(), schema.clone(), name.clone())) {
+            Some(appender) if !appender.tx.is_closed() => {
+                guard.insert((database, schema, name), appender.clone());
+                Ok::<_, AdapterError>(appender)
+            }
+            // We don't have a valid appender, so we need to get one.
+            //
+            // Note: we hold the lock while we acquire and appender to prevent a dogpile.
+            _ => {
+                tracing::info!(?database, ?schema, ?name, "fetching webhook appender");
+                adapter_client.metrics().webhook_get_appender.inc();
+
+                // Acquire and cache a new appender.
+                let appender = adapter_client
+                    .get_webhook_appender(database.clone(), schema.clone(), name.clone())
+                    .await?;
+
+                guard.insert((database, schema, name), appender.clone());
+
+                Ok(appender)
+            }
+        }
+    }
```

Comment:
Regarding the `tx.append(vec![(row, 1)]).await?;` line below: should everything between the `let AppendWebhookResponse` and the `tx.append` all be in a Retry loop? That way if any race condition happen (valid appender fetched, rename occurs, .append fails) we go all the way back up and try for another appender and append to it. If getting a new appender fails (because of a rename, then we can return a useful error (webhook does not exist or whatever). If getting a new appender succeeds (because of a swap), then the append will succeed. Would that remove one entire error class? I would really like to avoid returning a temporary "please retry" error if possible, and here seems like the place to put the retry loop.

Reply:
(Clearly the retry loop would need to know which errors are retryable and which are not.)

Reply:
This makes sense to me! Updated to move the append logic into its own function, and then wrapped it in an `mz_ore::retry::Retry`, where we try at most 2 times. Maybe only 2 is a bit limiting, but we realistically should never have to retry more than once for this race condition.